util-cache/src/main/scala/sbt/util/ActionCache.scala

```diff
@@ -0,0 +1,70 @@
+package sbt.util
+
+import scala.reflect.ClassTag
+import scala.annotation.{ meta, StaticAnnotation }
+import sjsonnew.{ HashWriter, JsonFormat }
+import sjsonnew.support.murmurhash.Hasher
+import xsbti.VirtualFile
+import java.nio.file.Path
+import scala.quoted.{ Expr, FromExpr, ToExpr, Quotes }
+
+object ActionCache:
+  def cache[I: HashWriter, O: JsonFormat: ClassTag](
+      key: I,
+      otherInputs: Long,
+      tags: List[CacheLevelTag],
+  )(
+      action: I => (O, Seq[VirtualFile])
+  )(
+      config: BuildWideCacheConfiguration
+  ): ActionResult[O] =
+    val hashInput: Array[Long] = Array(otherInputs, Hasher.hashUnsafe[I](key))
+    val input = Digest(HashUtil.sha256HashStr(hashInput))
+    val store = config.store
+    val outputDirectory = config.outputDirectory
+    val result: Option[ActionResult[O]] = store.get[O](input)
+    result match
+      case Some(value) =>
+        store.syncBlobs(value.outputFiles, outputDirectory)
+        value // return the value
+      case None =>
+        val (newResult, outputs) = action(key)
+        val value = store.put[O](input, newResult, outputs)
+        store.syncBlobs(value.outputFiles, outputDirectory)
+        value
+end ActionCache
+
+class BuildWideCacheConfiguration(
+    val store: ActionCacheStore,
+    val outputDirectory: Path,
+):
+  override def toString(): String =
+    s"BuildWideCacheConfiguration(store = $store, outputDirectory = $outputDirectory)"
+end BuildWideCacheConfiguration
+
+@meta.getter
+class cacheLevel(
+    include: Array[CacheLevelTag],
+) extends StaticAnnotation
+
+enum CacheLevelTag:
+  case Local
+  case Remote
+end CacheLevelTag
```

Comment:
I think the granularity of remote caching might need careful thinking to avoid turning small CPU-bound tasks to a network-bound one. On the other hand, I'm hoping that most tasks can be cached locally safely. I can't think of concrete examples but things like various mappings (file to path mapping) could be a candidate for marking local-only.